src/command.rs

```diff
@@ -295,23 +296,64 @@ impl CargoCmd {
     ///
     /// - `cargo 3ds build` and other "build" commands will use their callbacks to build the final `.3dsx` file and link it.
     /// - `cargo 3ds new` and other generic commands will use their callbacks to make 3ds-specific changes to the environment.
-    pub fn run_callback(&self, messages: &[Message]) {
-        // Process the metadata only for commands that have it/use it
-        let config = if self.should_build_3dsx() {
-            eprintln!("Getting metadata");
+    pub fn run_callback(&self, messages: &[Message], metadata: &Metadata) {
+        let mut configs = Vec::new();
 
-            Some(get_metadata(messages))
-        } else {
-            None
-        };
+        // Process the metadata only for commands that have it/use it
+        if self.should_build_3dsx() {
+            configs = self.run_build_callbacks(messages, metadata);
+        }
 
-        // Run callback only for commands that use it
+        // For run + test, we can only run one of the targets. Error if more or less
+        // than one executable was built, otherwise run the callback for the first target.
         match self {
-            Self::Build(cmd) => cmd.callback(&config),
-            Self::Run(cmd) => cmd.callback(&config),
-            Self::Test(cmd) => cmd.callback(&config),
+            Self::Run(cmd) if configs.len() == 1 => cmd.callback(&configs.into_iter().next()),
+            Self::Test(cmd) if configs.len() == 1 => cmd.callback(&configs.into_iter().next()),
+            Self::Run(_) | Self::Test(_) => {
+                let names: Vec<_> = configs.into_iter().map(|c| c.name).collect();
+                eprintln!("Error: expected exactly one executable to run, got {names:?}");
+                process::exit(1);
+            }
+            // New is a special case where we always want to run its callback
             Self::New(cmd) => cmd.callback(),
-            _ => (),
+            _ => {}
+        }
+    }
+
+    /// Generate a .3dsx for every executable artifact within the workspace that
+    /// was built by the cargo command.
+    fn run_build_callbacks(&self, messages: &[Message], metadata: &Metadata) -> Vec<CTRConfig> {
+        let mut configs = Vec::new();
```

Comment:
Useless suggestion (please don't really consider this): using `messages.len()` to pre-allocate the space needed.
```suggestion
        let mut configs = Vec::with_capacity(messages.len());
```

In most cases, this will end up over-allocating, **BUT** if someone were to build hundreds of targets, it might actually help with avoiding reallocations. :man_shrugging:

_Behaving like PCs have 1MHz DRAMs in the modern day._

Reply:
Hmm... your point about the performance not being super important is fair, but we might actually be able to do better than `messages.len()` here. The `Metadata` we already have at this point should allow us to count the number of possible artifacts, I think. Let me see what I can come up with...